### PR TITLE
AzureCliCredential: add support for `expires_on` field.

### LINKED ISSUE #5116: [Identity] Evaluate expires_on field in AzureCliCredential
As of [Azure CLI v2.54.0](https://github.com/Azure/azure-cli/milestone/135), the `az account get-access-token` command returns a new `expires_on` field containing a POSIX timestamp. This behavior is a departure from previous versions, in which the local datetime was returned in an `expiresOn` field. Update `AzureCliCredential` to also consider the new `expires_on` field.

Related links:
- https://github.com/Azure/azure-sdk-for-cpp/pull/5076#issuecomment-1784680918
- https://github.com/Azure/azure-cli/pull/27476

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #5116.

It works the following way - if `expires_on` (the new one) property is present, it gets parsed before `ExpiresOn` (the old one) property.

This PR is not a replacement for #5179, which mitigates the issue if the user has older version of Azure CLI installed. We can't expect users to always have the latest installed. The one without time zone info has been there first, for many years, and only now they are adding the `expires_on` property.

#5179 is good enough, but not ideal - once a year, one day in fall, when it is close to moving from DST, the token will be off for one hour, and then everything should work, but an extra token request will be sent when the auth policy will supply a token which was obtained when the DST was on, but the service will reject it, and then the policy will tell the credential to get a new token, and everything will work once again.
And one day in spring, when the clock moves to DST, the tokens obtained will have 1 hour shorter expiration than what they actually are, and some extra requests to refresh tokens will be made.

--
UWP CI is currently blocked until https://github.com/microsoft/vcpkg/issues/35279 is fixed / https://github.com/microsoft/vcpkg/pull/35116 is merged (#5181 would unblock)

### CODE REVIEW

**Comment by ahsonkhan** [comment]
Let's wait on this until we have a conclusion on Az Cli version support.

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp [hunk 1]
```diff
@@ -872,7 +872,7 @@ TEST(TokenCredentialImpl, Diagnosability)
           "}",
           "TokenForAccessing",
           "TokenExpiresInSeconds",
-          {}));
+          ""));
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:875]: Why are we changing this?

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp [hunk 2]
```diff
@@ -945,6 +945,52 @@ TEST(TokenCredentialImpl, Diagnosability)
...
   }
 
+  // Token is ok, relative expiration can't be parsed, two absolute expiration property names were
+  // provided, none of them can be parsed.
+  {
+    LogMsgVec log;
+    Logger::SetListener([&](auto lvl, auto msg) { log.push_back(std::make_pair(lvl, msg)); });
+
+    auto exceptionThrown = false;
+    try
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:955]: I don't understand what this test is trying to validate along with what this comment is trying to say. Can you please clarify.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 3]
```diff
@@ -221,6 +221,46 @@ TEST(AzureCliCredential, ExpiresIn)
   EXPECT_LE(token.ExpiresOn, timestampAfter + std::chrono::seconds(30));
 }
 
+TEST(AzureCliCredential, ExpiresOnUnixTimestampInt)
+{
+  constexpr auto Token = "{\"accessToken\":\"ABCDEFGHIJKLMNOPQRSTUVWXYZ\","
+                         "\"expiresOn\":\"2022-08-24 00:43:08.000000\","
+                         "\"expires_on\":1700692424}";
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:228]: So, if `expires_on` disagrees with `expiresOn`, the value of `expires_on` wins?

File: sdk/identity/azure-identity/src/token_credential_impl.cpp [hunk 4]
```diff
@@ -274,77 +277,80 @@ AccessToken TokenCredentialImpl::ParseToken(
     }
   }
 
-  if (expiresOnPropertyName.empty())
+  if (expiresOnPropertyNames.front().empty())
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/token_credential_impl.cpp:280]: We are picking the first one here to check. Why? It's a bit brittle to rely on the order in which the expiresOn property names vector is constructed. If we must rely on the order, there should be a comment at construction time highlighting that dependnecy.

File: sdk/identity/azure-identity/src/token_credential_impl.cpp [hunk 5]
```diff
@@ -274,77 +277,80 @@ AccessToken TokenCredentialImpl::ParseToken(
...
+            // parse() has thrown, we may throw later.
+          }
         }
       }
     }
   }
 
   ThrowJsonPropertyError(
-      expiresOnPropertyName,
+      expiresOnPropertyNames.back(),
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/token_credential_impl.cpp:349]: What happens if the vector is empty, or is it a precondition of this function that it never will be (or at least never should be)? Do we always expect the last one to be the one that failed at this point?

File: sdk/identity/azure-identity/src/token_credential_impl.cpp (hunk 4 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/token_credential_impl.cpp:280]: This is private, so I thought we could be less vandal-proof about the contract. But let me to update it.

File: sdk/identity/azure-identity/src/token_credential_impl.cpp (hunk 5 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/token_credential_impl.cpp:349]: Line 280 would throw and we will not get here.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp (hunk 3 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:228]: Yes. If it is present, we read from `expires_on`, and if we fail, we fall back to `ExpiresOn`.

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp (hunk 1 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:875]: `{}` could mean "empty vector" and "empty string". The new syntax unambiguously specifies it as "empty string".

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp (hunk 2 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:955]: I don't understand... The line above says it, as well as you can see from JSON below.
1. Token is ok.
2. Relative expiration ('TokenExpiresInSeconds') is not ok (`null`),
3. Absolute expiration (two properties - `TokenExpiresAtTime` and `token_expires_at_time`) are not ok (both are `null`).

The test verifies that we print the log message that involves the names of both absolute expiration properties:
```
Identity: TokenCredentialImpl::ParseToken():
Please report an issue with the following details:
Token JSON: Access token property ('TokenForAccessing'): string.length=11,
relative expiration property ('TokenExpiresInSeconds'): null,
absolute expiration property ('token_expires_at_time'): null,
absolute expiration property ('TokenExpiresAtTime'): null,
and there are no other properties.
```

Previously, the message would include only one absolute expiration property (`TokenExpiresAtTime`), now we check that both are printed.

Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp]: I put more comments for the test code.

File: sdk/identity/azure-identity/src/token_credential_impl.cpp (hunk 4 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/token_credential_impl.cpp]: I updated the code, now we check the entire vector for non-empty items.

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp (hunk 1 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp]: I updated the code to use `{}`, yet still not be ambiguous about its meaning (it won't compile otherwise) due to ambiguity.

**Comment by antkmsft** [comment]
> Let's wait on this until we have a conclusion on Az Cli version support.

Sure.
When will we know?
BTW, if the decision is to require the newer azure CLI version, I would want to talk to them, because there is no reason to drop the support entirely - we perfectly can parse timestamps in the old format - especially with #5179 we could do that, and it would be excessive in getting tokens twice a year in the locales that move to DST and back.

File: sdk/identity/azure-identity/src/azure_cli_credential.cpp [hunk 6]
```diff
@@ -154,7 +154,10 @@ AccessToken AzureCliCredential::GetToken(
       try
       {
         return TokenCredentialImpl::ParseToken(
-            azCliResult, "accessToken", "expiresIn", "expiresOn");
+            azCliResult,
+            "accessToken",
+            "expiresIn",
+            std::vector<std::string>{"expires_on", "expiresOn"});
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/azure_cli_credential.cpp:160]: The order of these two strings matter, correct?

We may be generalizing the approach prematurely by using a vector of strings, and it's introducing complexity.

I don't expect arbitrary set of strings to be passed in here, so we could add an extra parameter instead and simplify the parsing logic (no loops iterating through vectors and order/length checks).

File: sdk/identity/azure-identity/src/token_credential_impl.cpp [hunk 7]
```diff
@@ -274,77 +279,87 @@ AccessToken TokenCredentialImpl::ParseToken(
...
+  std::vector<std::string> nonEmptyExpiresOnPropertyNames;
+  std::copy_if(
+      expiresOnPropertyNames.begin(),
+      expiresOnPropertyNames.end(),
+      std::back_inserter(nonEmptyExpiresOnPropertyNames),
+      [](auto const& propertyName) { return !propertyName.empty(); });
+
+  if (nonEmptyExpiresOnPropertyNames.empty())
   {
     // 'expires_in' is undefined, 'expires_on' is not expected.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/token_credential_impl.cpp:291]: Does this comment need updating? I don't understand what it's saying.

**Comment by ahsonkhan** [comment]
> When will we know?
BTW, if the decision is to require the newer azure CLI version, I would want to talk to them, because there is no reason to drop the support entirely - we perfectly can parse timestamps in the old format - especially with https://github.com/Azure/azure-sdk-for-cpp/pull/5179 we could do that, and it would be excessive in getting tokens twice a year in the locales that move to DST and back.

I discussed this today and shared the conclusion on the issue:
https://github.com/Azure/azure-sdk-for-cpp/issues/5116#issuecomment-1832731537

File: sdk/identity/azure-identity/src/azure_cli_credential.cpp (hunk 6 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/azure_cli_credential.cpp:160]: The order of parameters matters - the vector is parsed in order, the items are defined.
Vector is readable, and we don't need to handle the cases if the first parameter is empty, but the second one is set. Even if we had two parameters, the logic for applying them is exactly the same, so I would create an array inside the function and iterate over it.

I think vector parameter is better.

File: sdk/identity/azure-identity/src/token_credential_impl.cpp (hunk 7 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/token_credential_impl.cpp:291]: `'expires_in'` is relative expiration, `'expires_on'` is absolute expiration. The comment is still valid.

File: sdk/identity/azure-identity/src/azure_cli_credential.cpp (hunk 6 above)
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/azure_cli_credential.cpp:160]: Please add a comment here, that the order matters. We want to check expires_on first, and only if that fails, fallback to the older expiresOn.

It would also be good to mention in a comment that parsing `expiresOn` field is only needed for versions of `Az CLI` older than v.2.54.0. Any new version will emit the new `expires_on` field.

This way, the source will explain **why** we are doing this, to our future selves :)

File: sdk/identity/azure-identity/src/azure_cli_credential.cpp [hunk 8]
```diff
@@ -153,8 +153,21 @@ AccessToken AzureCliCredential::GetToken(
...
       try
       {
+        // The order of elements in the vector below does matter - the code tries to find them
+        // consequently, and if finding the first one succeeds, we would not attempt to parse the
+        // second one. That is important, because the newer Azure CLI versions do have the new
+        // 'expires_on' field, which is not affected by time zone changes. The 'ExpiresOn' field was
+        // the only field that was present in the older versions, and it had problems, because it
+        // was a local timestamp without the time zone information.
+        // So, if only the 'expires_on' is available, we try to use it, and only if it is not
+        // available, we fall back to trying to get the value via 'ExpiresOn', which we also now are
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/azure_cli_credential.cpp:163]: The old field is "expiresOn" with that casing, to avoid confusion with the C++ `AccessToken` field.
```suggestion
        // 'expires_on' field, which is not affected by time zone changes. The 'expiresOn' field was
        // the only field that was present in the older versions, and it had problems, because it
        // was a local timestamp without the time zone information.
        // So, if only the 'expires_on' is available, we try to use it, and only if it is not
        // available, we fall back to trying to get the value via 'expiresOn', which we also now are
```

File: sdk/identity/azure-identity/src/private/token_credential_impl.hpp [hunk 9]
```diff
@@ -59,6 +59,31 @@ namespace Azure { namespace Identity { namespace _detail {
...
+    /**
+     * @brief Parses JSON that contains access token and its expiration.
+     *
+     * @param jsonString String with a JSON object to parse.
+     * @param accessTokenPropertyName Name of a property in the JSON object that represents access
+     * token.
+     * @param expiresInPropertyName Name of a property in the JSON object that represents token
+     * expiration in number of seconds from now.
+     * @param expiresOnPropertyNames Names of properties in the JSON object that represent token
+     * expiration as absolute date-time stamp. Can be empty, in which case no attempt to parse the
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/private/token_credential_impl.hpp:71]: Do all strings in the vector supposed to represent "expiration as absolute date-time stamp"?

File: sdk/identity/azure-identity/src/private/token_credential_impl.hpp [hunk 10]
```diff
@@ -59,6 +59,31 @@ namespace Azure { namespace Identity { namespace _detail {
...
+     * corresponding property will be made. Empty string elements will be ignored.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/private/token_credential_impl.hpp:72]: > Can be empty, in which case no attempt to parse the corresponding property will be made.

Is that comment true? I thought if the `expiresOnPropertyNames` vector was empty, this `ParseToken` function would throw.

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp [hunk 11]
```diff
@@ -894,7 +894,6 @@ TEST(TokenCredentialImpl, Diagnosability)
         "Please report an issue with the following details:\n"
         "Token JSON: Access token property ('TokenForAccessing'): string.length=11, "
         "relative expiration property ('TokenExpiresInSeconds'): \"one\", "
-        "absolute expiration property (''): undefined, "
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:897]: What's causing this test string to change? I wouldn't have expected this source change requiring a test update like this.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 12]
```diff
@@ -221,6 +221,56 @@ TEST(AzureCliCredential, ExpiresIn)
   EXPECT_LE(token.ExpiresOn, timestampAfter + std::chrono::seconds(30));
 }
 
+TEST(AzureCliCredential, ExpiresOnUnixTimestampInt)
+{
+  // 'expires_on' is 1700692424, which is a Unix timestamp of a date in 2023.
+  // 'ExpiresOn' is a date in 2022.
+  // The test checks that when both are present, 'expires_on' value (2023) is taken,
+  // and not that of 'ExpiresOn'.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:229]:
```suggestion
  // 'expiresOn' is a date in 2022.
  // The test checks that when both are present, 'expires_on' value (2023) is taken,
  // and not that of 'expiresOn'.
```

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 13]
```diff
@@ -221,6 +221,56 @@ TEST(AzureCliCredential, ExpiresIn)
...
+      token.ExpiresOn,
+      DateTime::Parse("2023-11-22T22:33:44.000000Z", DateTime::DateFormat::Rfc3339));
+}
+
+TEST(AzureCliCredential, ExpiresOnUnixTimestampString)
+{
+  // 'expires_on' is 1700692424, which is a Unix timestamp of a date in 2023.
+  // 'ExpiresOn' is a date in 2022.
+  // The test checks that when both are present, 'expires_on' value (2023) is taken,
+  // and not that of 'ExpiresOn'.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:253]:
```suggestion
  // 'expires_on' is 1700692424, which is a Unix timestamp of a date in 2023.
  // 'expiresOn' is a date in 2022.
  // The test checks that when both are present, 'expires_on' value (2023) is taken,
  // and not that of 'expiresOn'.
```

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 14]
```diff
@@ -221,6 +221,56 @@ TEST(AzureCliCredential, ExpiresIn)
...
+  // The test is similar to the one above, but the Unix timestamp is represented as string
+  // containing an integer.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:255]: Great!

Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:255]: Could the two tests be parameterized on the `Token` string with expires_on as a JSON number and as a JSON string, and hence avoid duplicating the test logic? The rest of the test source seems identical.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 15]
```diff
@@ -221,6 +221,56 @@ TEST(AzureCliCredential, ExpiresIn)
   EXPECT_LE(token.ExpiresOn, timestampAfter + std::chrono::seconds(30));
 }
 
+TEST(AzureCliCredential, ExpiresOnUnixTimestampInt)
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:224]: Can we add a test that results in an empty vector of strings being passed to `ParseToken` and also one where the vector of strings has elements that are `""`?

**Review by ahsonkhan — APPROVED** [review]
Otherwise, LGTM.

File: sdk/identity/azure-identity/src/private/token_credential_impl.hpp (hunk 9 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/private/token_credential_impl.hpp:71]: Yes, even the vector is called `expiresOnPropertyNames`. I.e. not `expiresInPropertyNames`.
"Expires In" = relative,
"Expires On"/"Expires At"  =absolute.

File: sdk/identity/azure-identity/src/private/token_credential_impl.hpp (hunk 10 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/private/token_credential_impl.hpp:72]: The comment is true. The function is internal. Your implementation does not expect any property to be there - you pass an empty value. It's not really a user input or server response, nor is it a user-facting functionality to throw exceptions. Function does what it is asked to. Empty value = the calling code does not expect one.

File: sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp (hunk 11 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/token_credential_impl_test.cpp:897]: These lines on the right: https://github.com/Azure/azure-sdk-for-cpp/pull/5180/files#diff-cc522e423ed15512dfbcccd5a21748c22c948f10b8b779f8201718aed16f98fcR476-R498
It is ok, though. Maybe even better You still can tell from the logs what was going on.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp (hunk 14 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:255]: I can find a way to parameterize a test with different values of same type, but not with the different values of different types.

Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:255]: The type is a JSON string in both cases.

You could parameterize on the whole `Token` string.

In one case it will be `"\"expires_on\": \"1700692424\""` and the other would be `"\"expires_on\": 1700692424"`

```text
  constexpr auto Token = "{\"accessToken\":\"ABCDEFGHIJKLMNOPQRSTUVWXYZ\","
                         "\"expiresOn\":\"2022-08-24 00:43:08.000000\"," // <-- 2022
                         "\"expires_on\":1700692424}"; // <-- 2023
```
